src/generic/tet_mesh.h

```diff
@@ -408,6 +408,192 @@ namespace oomph
+        }
+      }
+
+      // Store the offset of the current facet
+      unsigned current_offset = 0;
+
+      // This iterator is used for finding the index of a chosen vertex in
+      // 'vertices'
+      std::set<TetMeshVertex*>::iterator iterator;
+
+      // Get the offset, types and connectivity
+      for (auto individual_facet_pt : Facet_pt)
+      {
+        // Get the type of the current facet
+        n_vertices = individual_facet_pt->nvertex();
+
+        // The types of cell (polygon, tetra, etc) in the VTK/U file format are
+        // enumerated. The type of each facet is inputted using its
+        // corresponding enumeration into the 'types' section of the .vtu file.
+
+        // The relevant types of cells and enumerations are listed below.
+        // VTK_TRIANGLE = 5
+        // VTK_QUAD = 9
+        // VTK_POLYGON = 7
+
+        // The full list of types with enumerations are listed on the following
+        // site: https://vtk.org/wp-content/uploads/2015/04/file-formats.pdf
+
+        // Append the value corresponding to the current facet's type.
+        if (n_vertices == 3)
+        {
+          types.push_back(5);
+        }
+        else if (n_vertices == 4)
+        {
+          types.push_back(9);
+        }
+        else
+        {
+          types.push_back(7);
+        }
+
+        // Find the connectivity of this facet, the 'connectivity' vector
+        // consists of indices specifying the vertices that connect to create
+        // facets.
+        for (unsigned j = 0; j < n_vertices; j++)
+        {
+          iterator = vertices.find(individual_facet_pt->vertex_pt(j));
+          connectivity.push_back(std::distance(vertices.begin(), iterator));
+        }
+
+        // The vector 'offset' specifies which consecutive elements of
+        // 'connectivity' refer to a single facet.
+        current_offset += n_vertices;
+        offsets.push_back(current_offset);
+      }
+
+      //========================================================================
+      // Output to file
+      //========================================================================
+
+      // File Declaration
+      //-----------------
+
+      // Insert the necessary lines plus the number of vertices and cells/facets
+      // The number of facets is equal to the size of 'offsets'
+      outfile << "<VTKFile type=\"UnstructuredGrid\" version=\"0.1\" "
+                 "byte_order=\"BigEndian\">\n"
+              << "<UnstructuredGrid>\n"
+              << "<Piece NumberOfPoints=\"" << vertices.size()
+              << "\" NumberOfCells=\"" << offsets.size() << "\">" << std::endl;
+
+      // Vertices
+      //---------
+
+      outfile << "<Points>\n"
+              << "<DataArray type=\"Float32\" NumberOfComponents=\"3\" "
+                 "format=\"ascii\">"
+              << std::endl;
+
+      // Output the coordinates of every distinct vertex
+      for (auto vertex : vertices)
```

Comment:
Make these for-range loops use const references where possible, e.g.

```cpp
for (const auto& vertex : vertices)
...
```

Reply:
Fixed these issues in another commit: https://github.com/oomph-lib/oomph-lib/pull/98/commits/d93f2d9a9066f2626ab1076da0cc5dc579970122